File: static/settingsNG.js
/**
 * settingsNG.js -- settings-cog modal, top-right of the top bar.
 *
 * Curated actions (update yt-dlp, launch Suno Vault) plus an "Addresses"
 * section for the machine-specific endpoints in configNG.NG_ADDRESS_SETTINGS
 * (Ollama URL, Hermes gateway URL/key/model/session, Tailscale hostname).
 * Addresses are fetched/saved via routes/settingsNG.py, which persists
 * them to a gitignored JSON file (configNG.NG_SETTINGS_FILE) -- edits
 * take effect immediately, no server restart. Deliberately not an
 * open-ended "run any command" panel; each action gets its own reviewed
 * backend endpoint.
 *
 * Self-contained: only touches its own DOM (#ng-settings-cog,
 * #ng-settings-modal and children), so it doesn't need to sit in the
 * WIRING_ORDER chain -- loaded after bootstrapWiringNG.js in
 * indexNG.html, same as any other standalone feature file.
 */

(function () {
  const cogBtn = document.getElementById("ng-settings-cog");
  const overlay = document.getElementById("ng-settings-modal");
  const closeBtn = document.getElementById("ng-settings-modal-close");
  const updateBtn = document.getElementById("ng-settings-update-ytdlp");
  const outputEl = document.getElementById("ng-settings-ytdlp-output");
  const launchSunoBtn = document.getElementById("ng-settings-launch-suno");
  const sunoOutputEl = document.getElementById("ng-settings-suno-output");
  const launchComfyBtn = document.getElementById("ng-settings-launch-comfyui");
  const comfyOutputEl = document.getElementById("ng-settings-comfyui-output");
  const addressesEl = document.getElementById("ng-settings-addresses");
  const saveAddressesBtn = document.getElementById("ng-settings-save-addresses");
  const addressesOutputEl = document.getElementById("ng-settings-addresses-output");

  const SOURCE_LABEL = {
    env: "env var",
    saved: "saved",
    default: "default",
  };

  function renderAddressField(setting) {
    const wrap = document.createElement("div");
    wrap.className = "ng-settings-field";

    const head = document.createElement("div");
    head.className = "ng-settings-field-head";
    const label = document.createElement("strong");
    label.textContent = setting.label;
    const source = document.createElement("span");
    source.className = "ng-settings-field-source";
    source.textContent = SOURCE_LABEL[setting.source] || setting.source;
    head.appendChild(label);
    head.appendChild(source);

    const desc = document.createElement("span");
    desc.className = "ng-settings-row-desc";
    desc.textContent = setting.description || "";

    const input = document.createElement("input");
    input.type = setting.secret ? "password" : "text";
    input.dataset.settingKey = setting.key;
    input.value = setting.value || "";
    input.placeholder = setting.default || "";
    if (setting.source === "env") {
      // An env var always wins over a saved override -- editing here
      // would silently do nothing, so disable it and say why.
      input.disabled = true;
      source.textContent += " (takes precedence, edit not possible here)";
    }

    wrap.appendChild(head);
    wrap.appendChild(desc);
    wrap.appendChild(input);
    return wrap;
  }

  async function loadAddresses() {
    addressesEl.textContent = "Loading...";
    try {
      const res = await fetch("/api/ng/settings/addresses");
      const data = await res.json();
      addressesEl.textContent = "";
      (data.settings || []).forEach((setting) => {
        addressesEl.appendChild(renderAddressField(setting));
      });
    } catch (e) {
      addressesEl.textContent = "Failed to load: " + e.message;
    }
  }

  function openModal() {
    overlay.style.display = "flex";
    loadAddresses();
  }

  function closeModal() {
    overlay.style.display = "none";
  }

  cogBtn.addEventListener("click", openModal);
  closeBtn.addEventListener("click", closeModal);
  overlay.addEventListener("click", (e) => {
    if (e.target === overlay) closeModal();
  });

  updateBtn.addEventListener("click", async () => {
    updateBtn.disabled = true;
    updateBtn.textContent = "Updating...";
    outputEl.style.display = "block";
    outputEl.textContent = "Running pip install --upgrade yt-dlp ...";

    try {
      const res = await fetch("/api/ng/settings/update-ytdlp", { method: "POST" });
      const data = await res.json();

      if (data.ok) {
        outputEl.textContent = "Success.\n\n" + (data.output || "");
      } else {
        outputEl.textContent = "Failed: " + (data.error || "unknown error") +
          (data.output ? "\n\n" + data.output : "");
      }
    } catch (e) {
      outputEl.textContent = "Failed: " + e.message;
    } finally {
      updateBtn.disabled = false;
      updateBtn.textContent = "Update";
    }
  });

  launchSunoBtn.addEventListener("click", async () => {
    launchSunoBtn.disabled = true;
    launchSunoBtn.textContent = "Launching...";
    sunoOutputEl.style.display = "block";
    sunoOutputEl.textContent = "Starting Suno Vault server...";

    try {
      const res = await fetch("/api/ng/settings/launch-suno", { method: "POST" });
      const data = await res.json();

      if (data.ok) {
        sunoOutputEl.textContent =
          (data.already_running ? "Already running at " : "Started at ") +
          data.url;
        window.open(data.url, "_blank");
      } else {
        sunoOutputEl.textContent = "Failed: " + (data.error || "unknown error");
      }
    } catch (e) {
      sunoOutputEl.textContent = "Failed: " + e.message;
    } finally {
      launchSunoBtn.disabled = false;
      launchSunoBtn.textContent = "Launch";
    }
  });

  launchComfyBtn.addEventListener("click", async () => {
    launchComfyBtn.disabled = true;
    launchComfyBtn.textContent = "Launching...";
    comfyOutputEl.style.display = "block";
    comfyOutputEl.textContent = "Starting ComfyUI server (can take a while -- it's loading torch)...";

    try {
      const res = await fetch("/api/ng/settings/launch-comfyui", { method: "POST" });
      const data = await res.json();

      if (data.ok) {
        // No window.open here (unlike Suno above) -- the ComfyUI tab talks
        // to this server over its own API, nothing needs to navigate
        // anywhere, and popping an external origin is exactly what forced
        // a kill-and-restart of the installed iPad app after Suno's launch.
        comfyOutputEl.textContent =
          (data.already_running ? "Already running at " : "Started at ") +
          data.url +
          (data.logPath ? "\nLog: tail -f " + data.logPath : "");
      } else {
        comfyOutputEl.textContent = "Failed: " + (data.error || "unknown error");
      }
    } catch (e) {
      comfyOutputEl.textContent = "Failed: " + e.message;
    } finally {
      launchComfyBtn.disabled = false;
      launchComfyBtn.textContent = "Launch";
    }
  });

  saveAddressesBtn.addEventListener("click", async () => {
    const values = {};
    addressesEl.querySelectorAll("input[data-setting-key]").forEach((input) => {
      if (!input.disabled) values[input.dataset.settingKey] = input.value;
    });

    saveAddressesBtn.disabled = true;
    saveAddressesBtn.textContent = "Saving...";
    addressesOutputEl.style.display = "block";
    addressesOutputEl.textContent = "Saving...";

    try {
      const res = await fetch("/api/ng/settings/addresses", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(values),
      });
      const data = await res.json();

      if (data.ok) {
        addressesOutputEl.textContent = "Saved. Takes effect immediately.";
        addressesEl.textContent = "";
        (data.settings || []).forEach((setting) => {
          addressesEl.appendChild(renderAddressField(setting));
        });
      } else {
        addressesOutputEl.textContent = "Failed: " + (data.error || "unknown error");
      }
    } catch (e) {
      addressesOutputEl.textContent = "Failed: " + e.message;
    } finally {
      saveAddressesBtn.disabled = false;
      saveAddressesBtn.textContent = "Save";
    }
  });
})();

File: routes/settingsNG.py
"""immichRingNG -- settings-cog actions (venv maintenance, local launchers).

NG-only file, additive. Curated single endpoints rather than an open-
ended "run any command" one -- per the design note, this is meant to
grow with a reviewed list of specific actions, not become a shell.
"""

import os
import socket
import subprocess
import sys
import time
from urllib.parse import urlparse

from flask import Blueprint, jsonify, request

from configNG import (
    COMFYUI_DIR,
    SUNO_DIR,
    SUNO_PORT,
    get_comfyui_base_url,
    get_ng_address_settings,
    get_tailscale_hostname,
    save_ng_address_settings,
)

settingsNG_bp = Blueprint("settingsNG", __name__)


@settingsNG_bp.route("/api/ng/settings/addresses", methods=["GET"])
def get_addresses_ng():
    """List every machine-specific address setting (Ollama/Hermes/Tailscale)
    with its current effective value and where that value came from --
    drives the "Addresses" section of the settings modal. Secret values
    (Hermes API key) are still sent as plain text here since this is a
    same-origin, single-user local app; the UI masks them on display."""
    return jsonify({"settings": get_ng_address_settings()})


@settingsNG_bp.route("/api/ng/settings/addresses", methods=["POST"])
def save_addresses_ng():
    """Persist edited address settings to configNG.NG_SETTINGS_FILE. Body:
    {key: value, ...} -- a blank value clears that key's override, falling
    back to its env var / hardcoded default. Takes effect immediately, no
    restart needed (configNG reads the file fresh on every call)."""
    body = request.get_json(silent=True) or {}
    if not isinstance(body, dict):
        return jsonify({"ok": False, "error": "body must be a JSON object"}), 400
    save_ng_address_settings(body)
    return jsonify({"ok": True, "settings": get_ng_address_settings()})


@settingsNG_bp.route("/api/ng/settings/update-ytdlp", methods=["POST"])
def update_ytdlp_ng():
    """Run `pip install --upgrade yt-dlp` in the current venv (the same
    interpreter running this Flask process) and report back pass/fail
    plus the tail of pip's output for visibility."""
    try:
        result = subprocess.run(
            [sys.executable, "-m", "pip", "install", "--upgrade", "yt-dlp"],
            capture_output=True,
            text=True,
            timeout=120,
        )
    except subprocess.TimeoutExpired:
        return jsonify({
            "ok": False,
            "error": "Timed out after 120s waiting for pip.",
        }), 504
    except Exception as e:
        return jsonify({"ok": False, "error": str(e)}), 500

    output = (result.stdout or "") + (result.stderr or "")
    tail = "\n".join(output.strip().splitlines()[-15:])

    if result.returncode != 0:
        return jsonify({
            "ok": False,
            "error": f"pip exited with code {result.returncode}",
            "output": tail,
        }), 500

    return jsonify({"ok": True, "output": tail})


def _port_open(host, port, timeout=0.5):
    try:
        with socket.create_connection((host, port), timeout=timeout):
            return True
    except OSError:
        return False


@settingsNG_bp.route("/api/ng/settings/launch-suno", methods=["POST"])
def launch_suno_ng():
    """Start john's Suno Vault server (~/Music/mysunodb, its own repo and
    venv) if it isn't already running, and hand back its Tailscale URL.
    Started detached (start_new_session) so it keeps running after this
    request -- and after this Flask process -- exits."""
    url = f"http://{get_tailscale_hostname()}:{SUNO_PORT}/"

    if _port_open("127.0.0.1", SUNO_PORT):
        return jsonify({"ok": True, "already_running": True, "url": url})

    if not os.path.isdir(SUNO_DIR):
        return jsonify({
            "ok": False,
            "error": f"Suno Vault directory not found: {SUNO_DIR}",
        }), 500

    venv_python = os.path.join(SUNO_DIR, ".venv", "bin", "python3")
    python_bin = venv_python if os.path.exists(venv_python) else sys.executable

    try:
        subprocess.Popen(
            [python_bin, "server.py"],
            cwd=SUNO_DIR,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            start_new_session=True,
        )
    except Exception as e:
        return jsonify({"ok": False, "error": str(e)}), 500

    for _ in range(20):
        if _port_open("127.0.0.1", SUNO_PORT):
            return jsonify({"ok": True, "already_running": False, "url": url})
        time.sleep(0.25)

    return jsonify({
        "ok": False,
        "error": "Server didn't come up within 5s -- check it manually.",
    }), 500


@settingsNG_bp.route("/api/ng/settings/launch-comfyui", methods=["POST"])
def launch_comfyui_ng():
    """Start the local ComfyUI server (/Volumes/AI/ComfyUI, its own venv)
    if it isn't already running. Host/port come from the comfyui_base_url
    address setting rather than a hardcoded constant (unlike Suno) so this
    stays in sync if that setting is ever repointed -- only makes sense to
    call this when that URL actually names this machine, though; there's
    no check for that here.

    Deliberately doesn't open/navigate anywhere on success (contrast
    launch_suno_ng's window.open) -- ComfyUI's own graph UI isn't the
    point, the ComfyUI tab's own API calls to this server are, and
    window.open(..., "_blank") on an iPad PWA hands the external origin
    to Safari in a way that has left the installed app needing a force-
    quit to recover (John's report from the Suno launch button)."""
    base_url = get_comfyui_base_url()
    parsed = urlparse(base_url)
    port = parsed.port or 8188

    already_log = os.path.join(COMFYUI_DIR, "comfyui_launch.log")
    if _port_open("127.0.0.1", port):
        # If it's already running, this may well be an older instance
        # started before this log-file redirect existed (or started some
        # other way entirely) -- only claim a log path if one's actually
        # there to tail, rather than pointing at a file with nothing in it.
        resp = {"ok": True, "already_running": True, "url": base_url}
        if os.path.isfile(already_log):
            resp["logPath"] = already_log
        return jsonify(resp)

    if not os.path.isdir(COMFYUI_DIR):
        return jsonify({
            "ok": False,
            "error": f"ComfyUI directory not found: {COMFYUI_DIR}",
        }), 500

    venv_python = os.path.join(COMFYUI_DIR, "venv", "bin", "python3")
    python_bin = venv_python if os.path.exists(venv_python) else sys.executable

    # Truncated fresh on every launch -- this is "this run's log", not an
    # accumulating history, so `tail -f` always shows what the currently
    # running process is actually doing. Previously stdout/stderr went to
    # DEVNULL, so a launch's console output (custom-node errors, sampler
    # progress, anything short of ComfyUI's own web UI) was unrecoverable
    # -- John asked where to find it and there was nothing to point at.
    log_path = os.path.join(COMFYUI_DIR, "comfyui_launch.log")
    try:
        with open(log_path, "w") as log_file:
            subprocess.Popen(
                [python_bin, "main.py", "--listen", "--port", str(port)],
                cwd=COMFYUI_DIR,
                stdout=log_file,
                stderr=subprocess.STDOUT,
                start_new_session=True,
            )
        # Popen dup()s the fd for the child at spawn time, so closing our
        # own handle here (the `with` block exiting) doesn't affect the
        # now-independent detached process's writes to it.
    except Exception as e:
        return jsonify({"ok": False, "error": str(e)}), 500

    # ComfyUI is a much heavier startup than Suno's plain server (loads
    # torch, scans custom nodes) -- give it real time before giving up.
    for _ in range(60):
        if _port_open("127.0.0.1", port):
            return jsonify({"ok": True, "already_running": False, "url": base_url, "logPath": log_path})
        time.sleep(0.5)

    return jsonify({
        "ok": False,
        "error": "Server didn't come up within 30s -- check it manually.",
    }), 500
